src/metabase/util/memoize/impl/bounded.cljc

```diff
@@ -0,0 +1,32 @@
+(ns metabase.util.memoize.impl.bounded
+  (:require
+   [metabase.util.log :as log]
+   #?(:clj  [clojure.core.cache :as cache]
+      :cljs [cljs.cache :as cache])))
+
+(cache/defcache BoundedCache [cache threshold]
+  cache/CacheProtocol
+  (lookup [_ item] (get cache item))
+  (lookup [_ item not-found] (get cache item not-found))
+  (has?   [_ item] (contains? cache item))
+  (hit    [this _item] this)
+  (miss   [_ item result]
+    (if (< (count cache) threshold)
+      (BoundedCache. (assoc cache item result) threshold)
+      ;; It's too big! Throw away the original cache and start over!
+      (do
+        (log/infof "BoundedCache threshold (%d) exceeded - runaway cache? threshold too small?" threshold)
```

Comment:
Doesn't this deserve to be a warning? We don't expect this to happen.

Reply:
I suppose so. It seems a bit trivial for that level, but we would want to see it reported by any user who was seeing it in the wild.